nexus/db-queries/src/db/datastore/inventory.rs

```diff
@@ -1285,73 +1264,115 @@ impl DataStore {
 
         let errors: Vec<String> = {
             use db::schema::inv_collection_error::dsl;
-            dsl::inv_collection_error
+            let mut errors = Vec::new();
+            let mut paginator = Paginator::new(batch_size);
+            while let Some(p) = paginator.next() {
+                let batch = paginated(
+                    dsl::inv_collection_error,
+                    dsl::idx,
+                    &p.current_pagparams(),
+                )
                 .filter(dsl::inv_collection_id.eq(id))
                 .order_by(dsl::idx)
-                .limit(sql_limit)
                 .select(InvCollectionError::as_select())
                 .load_async(&*conn)
                 .await
-                .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?
-                .into_iter()
-                .map(|e| e.message)
-                .collect()
+                .map_err(|e| {
+                    public_error_from_diesel(e, ErrorHandler::Server)
+                })?;
+                paginator =
+                    p.found_batch(&batch, &|row: &InvCollectionError| row.idx);
+                errors.extend(batch.into_iter().map(|e| e.message));
+            }
+            errors
         };
-        limit_reached = limit_reached || errors.len() == usize_limit;
 
         let sps: BTreeMap<_, _> = {
             use db::schema::inv_service_processor::dsl;
-            dsl::inv_service_processor
+
+            let mut sps = BTreeMap::new();
+
+            let mut paginator = Paginator::new(batch_size);
+            while let Some(p) = paginator.next() {
+                let batch = paginated(
+                    dsl::inv_service_processor,
+                    dsl::time_collected,
+                    &p.current_pagparams(),
+                )
                 .filter(dsl::inv_collection_id.eq(id))
-                .limit(sql_limit)
                 .select(InvServiceProcessor::as_select())
                 .load_async(&*conn)
                 .await
-                .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?
-                .into_iter()
-                .map(|sp_row| {
-                    let baseboard_id = sp_row.hw_baseboard_id;
+                .map_err(|e| {
+                    public_error_from_diesel(e, ErrorHandler::Server)
+                })?;
+                paginator = p.found_batch(&batch, &|row| row.time_collected);
```

Comment:
The key here really should be unique, but `time_collected` isn't.  How about using `hw_baseboard_id`?

Reply:
Ah, that's what I'd done initially, but when I read the documentation for `Paginator`, I got the impression that this was supposed to be how the rows were _ordered_. I'll change it to the unique ID here (and elsewhere).

Reply:
Ah, yeah, it should be both: we want an index that's ordered the way we want to paginate through them, and where each item can be uniquely identified by one the last value in the index.

Reply:
Got it, thanks. I wasn't sure if the UUIDs met the ordering requirement in this case, but based on your suggestion above I'm assuming they do?

Reply:
That's a good question.  If we take this example of `inv_service_processor.hw_baseboard_id`, [we have a composite primary key of `inv_collection_id` (which is part of the filter here) and then `hw_baseboard_id`](https://github.com/oxidecomputer/omicron/blob/5439ce8f391e1f2f1679378171c5a3d880f7ee37/schema/crdb/dbinit.sql#L2784), so yes, we'll have a unique index sorted first by `inv_service_processor`, then by `hw_baseboard_id`, which means we can (1) fetch any page of rows of `inv_service_processor` by going directly to the corresponding part of the index, and then (2) uniquely identify the last row on any page.  (By "page" here I mean the sense used in this code, not an OS page or filesystem page or database page or anything like that.)

We'd want to check that for all the other cases, too.  In some cases, it might not be the primary key, but there may be another unique index.  And if there isn't already a suitable unique index, we'll probably want to add one to the schema.  I'm happy to talk through this live if that's useful, especially if we find we have to add an index.

Reply:
I think everything should be using the right keys for pagination as of 02f39d2ae8b1898cad45c6914f0d3f072dd2a3dd --- if I'm wrong about that, happy to chat about it either in review or synchronously!